easybuild/tools/modules.py

```diff
@@ -831,33 +834,54 @@ def available(self, mod_name=None):
 
         return correct_real_mods
 
+    def compile_cache(self, cache_fp):
+        """Compile Lmod cache file using luac."""
+        if build_option('compile_lmod_caches'):
+            # FIXME: ask Lmod to compile, to ensure that luac that matches the lua used by Lmod is used?
+            run_cmd("luac %s" % cache_fp)
+
     def update(self):
-        """Update after new modules were added."""
-        spider_cmd = os.path.join(os.path.dirname(self.cmd), 'spider')
-        cmd = [spider_cmd, '-o', 'moduleT', os.environ['MODULEPATH']]
-        self.log.debug("Running command '%s'..." % ' '.join(cmd))
+        """Update Lmod cache after new modules were added."""
+        cache_dir = build_option('update_lmod_caches')
 
-        proc = subprocess.Popen(cmd, stdout=PIPE, stderr=PIPE, env=os.environ)
-        (stdout, stderr) = proc.communicate()
+        if cache_dir is not None:
 
-        if stderr:
-            self.log.error("An error occured when running '%s': %s" % (' '.join(cmd), stderr))
+            for cache_type in ['moduleT', 'dbT', 'reverseMapT']:
 
-        if self.testing:
-            # don't actually update local cache when testing, just return the cache contents
-            return stdout
-        else:
-            try:
-                cache_filefn = os.path.join(os.path.expanduser('~'), '.lmod.d', '.cache', 'moduleT.lua')
-                self.log.debug("Updating Lmod spider cache %s with output from '%s'" % (cache_filefn, ' '.join(cmd)))
-                cache_dir = os.path.dirname(cache_filefn)
-                if not os.path.exists(cache_dir):
-                    mkdir(cache_dir, parents=True)
-                cache_file = open(cache_filefn, 'w')
-                cache_file.write(stdout)
-                cache_file.close()
-            except (IOError, OSError), err:
-                self.log.error("Failed to update Lmod spider cache %s: %s" % (cache_filefn, err))
+                # run spider to create cache contents
+                spider_cmd = os.path.join(os.path.dirname(self.cmd), 'spider')
+                cmd = [spider_cmd, '-o', cache_type, os.environ['MODULEPATH']]
+                self.log.debug("Running command '%s'..." % ' '.join(cmd))
+
+                # we can't use run_cmd, since we need to separate stdout from stderr
```

Comment:
we should start using the `run` commands that are provided by `vsc.utils.run`, and the function in there should be enhanced to allow spltiting stdout/stderr

that's a long-term effort though, since either we redefine `run_cmd` to use those, or we drop/deprecate `run_cmd` all together